vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,787 @@
+{
+  if (cn.examples.size() == 0) { return nullptr; }
+
+  float best_score = FLT_MAX;
+  emt_example* best_example = cn.examples[0].get();
+
+  // shuffle the examples to break ties randomly
+  emt_shuffle(cn.examples.begin(), cn.examples.end(), b._random_state.get());
+
+  for (auto const& example : cn.examples)
+  {
+    float score = scorer_predict(b, base, ex, *example);
+
+    if (score < best_score)
+    {
+      best_score = score;
+      best_example = example.get();
+    }
+  }
+
+  return best_example;
+}
+
+void node_predict(emt_tree& b, single_learner& base, emt_node& cn, emt_example& ex, VW::example& ec)
+{
+  auto closest_ex = node_pick(b, base, cn, ex);
+  ec.pred.multiclass = (closest_ex != nullptr) ? closest_ex->label : 0;
+  ec.loss = (ec.l.multi.label != ec.pred.multiclass) ? ec.weight : 0;
+}
+
+void predict(emt_tree& b, single_learner& base, VW::example& ec)
+{
+  b.all->ignore_some_linear = false;
+  emt_example ex(*b.all, &ec);
+
+  emt_node& cn = *tree_route(b, ex);
+  node_predict(b, base, cn, ex, ec);
+  tree_bound(b, &ex);
+}
+
+void learn(emt_tree& b, single_learner& base, VW::example& ec)
+{
+  b.all->ignore_some_linear = false;
+  auto ex = VW::make_unique<emt_example>(*b.all, &ec);
+
+  emt_node& cn = *tree_route(b, *ex);
+  scorer_learn(b, base, cn, *ex, ec.weight);
+  node_predict(b, base, cn, *ex, ec);  // vw learners predict and learn
+
+  tree_bound(b, ex.get());
+  node_insert(cn, std::move(ex));
+  node_split(b, cn);
+}
+
+void end_pass(emt_tree& b)
+{
+  *(b.all->trace_message) << "##### pass_time: " << static_cast<float>(clock() - b.begin) / CLOCKS_PER_SEC << std::endl;
+
+  b.begin = clock();
+}
+/////////////////end of eigen_memory_tree///////////////////
+////////////////////////////////////////////////////////////
+
+///////////////////Save & Load//////////////////////////////////////
+////////////////////////////////////////////////////////////////////
+void save_load_examples(emt_node& n, io_buf& model_file, bool& read, bool& text, std::stringstream& msg)
+{
+  WRITEITVAR(n.examples.size(), "n_examples", n_examples);
+
+  if (read)
+  {
+    for (uint32_t i = 0; i < n_examples; i++) { n.examples.push_back(VW::make_unique<emt_example>()); }
+  }
+
+  for (auto& e : n.examples)
+  {
+    for (auto& p : e->base)
+    {
+      WRITEIT(p.first, "base_index");
+      WRITEIT(p.second, "base_value");
+    }
+    for (auto& p : e->full)
+    {
+      WRITEIT(p.first, "full_index");
+      WRITEIT(p.second, "full_value");
+    }
+  }
+}
+
+void save_load_weights(emt_node& n, io_buf& model_file, bool& read, bool& text, std::stringstream& msg)
+{
+  for (auto& p : n.router_weights)
+  {
+    WRITEIT(p.first, "router_index");
+    WRITEIT(p.second, "router_value");
+  }
+}
+
+std::unique_ptr<emt_node> save_load_node(
+    emt_tree& b, std::unique_ptr<emt_node> n, io_buf& model_file, bool& read, bool& text, std::stringstream& msg)
+{
+  WRITEITVAR(!read && !n, "is_null", is_null);
```

Comment:
Essentially, you need to implement `read_model_field` and `write_model_field` for your new types to allow for serialization/deserialization.

Reply:
https://github.com/VowpalWabbit/vowpal_wabbit/blob/5ca2a7f54a9bf02794c1417b68439538d638555c/vowpalwabbit/core/src/reductions/epsilon_decay.cc#L274